### ISSUE TITLE
Fails with " unterminated quoted string"

### ISSUE THREAD

**Issue by bspiegel33** [PomDev2/wlanpoke#7]
Fresh install (v 0.8.4.1), no change to code and I get this error:

sh /etc/wwlanpoke/wlanpoke.sh &

wlanpoke.sh: eval: line 1: syntax error: unterminated quoted string
wlanpoke exiting

I assume there's some settings that I need to change?

**Comment by benfitzg** [comment]
Hi, I had this also. I tracked it down to a neighbour putting an apostrophe in their network name.

Run this on your radio:
`wpa_cli scan_results | grep -v -e interface -e signal`

If that shows a network id with a quote here is the patch:

Sorry this isn't a PR. 

```
diff --git a/wlanpoke.sh b/wlanpoke.sh
index acc8fe2..81a0847 100644
--- a/wlanpoke.sh
+++ b/wlanpoke.sh
@@ -692,7 +692,7 @@ GetStats () {

   # 0.5.2 esperimental: add lengthy 'wpa_cli scan_results' to see if the scanner is losing the AP.
   # results of PREVIOUS scan
-  LASTSTAT=$LASTSTAT" "`wpa_cli scan_results | grep -v -e interface -e signal`
+  LASTSTAT=$LASTSTAT" "`wpa_cli scan_results | grep -v -e interface -e signal | sed "s#'#\\\'#g"`

   # request a new scan -- if we dare -- don't want to disrupt the existing connection unnecessarily.
   # wpa_cli scan

```

This is just a cursory fix to stop the script complaining, not sure if it would work if it's your own SSID. I would say remove the quote from your SSID!

**Comment by bspiegel33** [comment]
Thanks!
My network is net.tone so I assume the issue is the "."
I'll try the fix.

**Comment by bspiegel33** [comment]
Here's the results of 
wpa_cli scan_results | grep -v -e interface -e signal
No quotes but some dashes and periods.

`
anita  -  2.4
Pune
net.tone.1
ShivAadi
ShivAadi-guest
ShivAadi
ShivAadi-guest
NETGEAR24
Aravind123-2.4
NAKO
NETGEAR33
TP-Link_Extender
`

**Comment by benfitzg** [comment]
Perhaps! I found this by running with debug on ash:
`ash -x wlanpoke.sh >output.txt  2>&1 &`
Then scp'd the file off and looked at the eval just before the error output, confirmed the presence (in my case) of a quote, then wrote that sed expression, which fixed my problem. You can probably find the offending eval. The problem with eval is it's evaluated as a separate expression so the line number error tells you nothing about where in wlanpoke.sh the error occurs.

**Comment by bspiegel33** [comment]
You were right.  I ran several times and when it failed, I discovered a neighbors wifi  "Stan's Wifi".  Your fix seems to be working. 

Many thanks.

**Comment by bspiegel33** [comment]
Hum, didn't work after all.  However, I replaced the sed with a grep -v "'" to eliminate my neighbors wifi and works.  Thanks so much for highlighting the issue.

**Comment by benfitzg** [comment]
Very happy to feed something back into this excellent project from PomDev2.

Would be good to know the root cause for your case so a patch can go in.

**Comment by PomDev2** [comment]
Thanks for all this. We would like to show all access points, since they may be implicated in the interference. Given your experiences, what is your best recommendation for an update to the script?

And thanks for the ash -x switch, which perhaps can be used to find a problem in which the script exits with a shell error after an extended period of time (many days). However, the script running under -x generates a lot of output, too much to write to the file system, so try nc'ing a pipe output to the server:

```
TCPPORT="1122"          # 1121 is for wlanpoke
TCPLOG=$SERVER			# replace this with ip address of ncat listening on 1122
WLPIPENAME="tester"
rm -f $WLPIPENAME
mkfifo $WLPIPENAME
while true; do cat $WLPIPENAME | nc -w 3 $TCPLOG $TCPPORT ; done &
ash -x /etc/wlanpoke/wlanpoke.sh -x -W quick -d /etc/wlanpoke/log/ >$WLPIPENAME 2>&1 &

```
The wlanpoke load with the tracing is 13%, pretty high, but the radio still works. We'll see what happens when the network fails! Let's hope the pipe fills until the connection next succeeds and empties the pipe.

**Comment by benfitzg** [comment]
In my case the wpa_cli scan_results immediately spat out the problematic id, so I didn't leave it running for more than a few seconds. it's true that if the offending network id only pops up infrequently, you might need to leave it running for a long time before you can catch it "in the act". Then streaming the logs makes more sense.

I think you just need to escape the offending chars so they don't interfere with eval. Because you are using eval that can be a bit of a pain as you have to escape the escape or the shell eats it.

**Comment by PomDev2** [comment]
Thanks for your very helpful input. It looks like your problem occurred in the CB_put() function called from GetStats (). Is that correct? As noted in the CB_put test function, single and double quotes are not supported. But your example shows that single quote support is actually required.
```
CB_put () {
  eval "a$CB_HEAD='$@'"    # shell error here with a single quote?
```
As noted in the script, the eval() function provides array-like functionality otherwise missing in ash, and is frequently used. Can you suggest a way to improve this assignment, either by escaping (which didn't work earlier???), substituting for the '\''?,  or some way other than eval to implement an array?
```
CB_put () {
  local args=$(echo $@ | sed "s/'/--/g")  # substitute any '\'' characters
  eval "a$CB_HEAD='$args'"
```
(For me "--" is easier to read and explain than "\'", although the latter is of course close to the original.)
Thanks for your help.

**Comment by benfitzg** [comment]
Yes it was in CB_put. That will fix the issue for a quote. I'm surprised a network ID allows such a character, which begs the question what else might be in there? I wonder if you should cover wider cases like sed 's/[^a-zA-Z-_]//g'. Up to you, I'm not an expert in this area.

**Comment by PomDev2** [comment]
This issue has been fixed and published starting with version 0.8.7.1 (now in the development branch).
